src/graph/immutable_graph.cc

```diff
@@ -20,6 +22,80 @@ bool binary_search(ForwardIt first, ForwardIt last, const T& value) {
   return (!(first == last) && !(value < *first));
 }
 
+ImmutableGraph::CSR::CSR(int64_t num_vertices, int64_t expected_num_edges):
+    indptr(num_vertices + 1), indices(expected_num_edges), edge_ids(expected_num_edges) {
+  indptr.resize(num_vertices + 1);
+}
+
+ImmutableGraph::CSR::CSR(IdArray indptr_arr, IdArray index_arr, IdArray edge_id_arr):
+    indptr(indptr_arr->shape[0]), indices(index_arr->shape[0]), edge_ids(index_arr->shape[0]) {
+  size_t num_vertices = indptr_arr->shape[0] - 1;
+  size_t num_edges = index_arr->shape[0];
+  const int64_t *indptr_data = static_cast<int64_t*>(indptr_arr->data);
+  const dgl_id_t *indices_data = static_cast<dgl_id_t*>(index_arr->data);
+  const dgl_id_t *edge_id_data = static_cast<dgl_id_t*>(edge_id_arr->data);
+  CHECK_EQ(indptr_data[0], 0);
+  CHECK_EQ(indptr_data[num_vertices], num_edges);
+  indptr.insert_back(indptr_data, num_vertices + 1);
+  indices.insert_back(indices_data, num_edges);
+  edge_ids.insert_back(edge_id_data, num_edges);
+}
+
+ImmutableGraph::CSR::CSR(IdArray indptr_arr, IdArray index_arr, IdArray edge_id_arr,
+                         const std::string &shared_mem_name) {
+#ifndef _WIN32
+  size_t num_vertices = indptr_arr->shape[0] - 1;
+  size_t num_edges = index_arr->shape[0];
+  CHECK_EQ(num_edges, edge_id_arr->shape[0]);
+  size_t file_size = (num_vertices + 1) * sizeof(int64_t) + num_edges * sizeof(dgl_id_t) * 2;
+
+  auto mem = std::make_shared<runtime::SharedMemory>(shared_mem_name);
+  auto ptr = mem->create_new(file_size);
+
+  int64_t *addr1 = static_cast<int64_t *>(ptr);
+  indptr.init(addr1, num_vertices + 1);
+  void *addr = addr1 + num_vertices + 1;
+  dgl_id_t *addr2 = static_cast<dgl_id_t *>(addr);
+  indices.init(addr2, num_edges);
+  addr = addr2 + num_edges;
+  dgl_id_t *addr3 = static_cast<dgl_id_t *>(addr);
+  edge_ids.init(addr3, num_edges);
+
+  const int64_t *indptr_data = static_cast<int64_t*>(indptr_arr->data);
+  const dgl_id_t *indices_data = static_cast<dgl_id_t*>(index_arr->data);
+  const dgl_id_t *edge_id_data = static_cast<dgl_id_t*>(edge_id_arr->data);
+  CHECK_EQ(indptr_data[0], 0);
+  CHECK_EQ(indptr_data[num_vertices], num_edges);
+  indptr.insert_back(indptr_data, num_vertices + 1);
+  indices.insert_back(indices_data, num_edges);
+  edge_ids.insert_back(edge_id_data, num_edges);
+  this->mem = mem;
+#else
+  LOG(FATAL) << "Windows doesn't support ImmutableGraph with shared memory";
```

Comment:
`Shared memory support not yet implemented`.

Windows does support shared memory.